BUG: Series rolling count ignores min_periods

fujiaxiang opened the pull request:
- [x] closes #26996
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

fujiaxiang committed:
BUG: Series rolling count ignores min_periods (GH26996)

WillAyd commented on pandas/core/window/rolling.py:
```diff
@@ -1185,14 +1185,16 @@ def count(self):
                 result,
                 window=window,
-                min_periods=0,
+                min_periods=min_periods,
```
What is the downside of just passing `self.min_periods` here? Would be more consistent with other places in the module

fujiaxiang replied:
hi Will, this is for backward compatibility purpose. If we set `min_periods=self.min_periods`, many (63 to be exact) existing test cases fail, which all assume `min_periods` is 0. 

Ping me if you think we should break this backward-compatibility in favor of consistency with other APIs.

WillAyd replied:
Are the tests wrong? Seems like we should be consistent

fujiaxiang replied:
I just realize the real reason behind this design (making ``min_periods=0``): because ``min_periods`` is interpreted as "minimum number of valid (i.e. not null) entries" rather than "minimum number of entries in the current window".

To illustrate what I mean:
```python
>>> s = pd.Series([1, 2, 3, np.nan, 4, 5, 6])
>>> s
0    1.0
1    2.0
2    3.0
3    NaN
4    4.0
5    5.0
6    6.0
dtype: float64

>>> s.rolling(3, min_periods=3).mean()  # note we have NaNs for iloc[3:6]
0    NaN
1    NaN
2    2.0
3    NaN
4    NaN
5    NaN
6    5.0
dtype: float64
```

Now, the exact point of `count` method is to count how many values are not NaNs, if we make this exactly the same as other APIs, i.e. by default `min_periods=None` which will then be assigned to be `min_periods = window` inside the functions, then we would give NA as long as there is at least one NA in the window, or full window length otherwise, like this:
```python
# This would be the result if we make rolling count behave the same as other APIs. 
# I feel this should not be the default behavior
>>> s.rolling(3).count()  
0    NaN
1    NaN
2    3.0
3    NaN
4    NaN
5    NaN
6    3.0

# This is what I think it should be
>>> s.rolling(3).count()
0    1.0
1    2.0
2    3.0
3    2.0
4    2.0
5    2.0
6    3.0

# However I do want to allow the user to specify min_periods 
# and make the function behave like the others
>>> s.rolling(3, min_periods=3).count()  
0    NaN
1    NaN
2    3.0
3    NaN
4    NaN
5    NaN
6    3.0
```
The last case is what this PR is trying to achieve.

WillAyd reviewed (changes requested):
Thanks for the PR. Looks generally good just a question

fujiaxiang committed:
Merge remote-tracking branch 'upstream/master' into series_rolling_count_ignores_min_periods

fujiaxiang committed:
ENH: handles min_periods argument in rolling.count (GH26996)

fujiaxiang commented:
As mentioned in my previous reply to a change request, I think we should properly handles `min_periods` when it is explicitly set by user, but the default behavior should remain the same, i.e. `min_periods=0` by default (it defaults to `None` in method signature but inside is set to 0). 

The reason is that this `count` method is meant to count the non-NAs values in the window, and we don't want to require all values in the window to be valid in order to produce a valid count, which is what other similar APIs do (e.g. rolling.mean sets `min_periods=window`, effectively requires all values in the window to be valid, otherwise produces `NaN`)

However, if you feel strongly we should make the default behavior exactly the same as other APIs like `rolling.mean`, let me know and I will update this PR. I have the code ready. The new code updates 63 existing test cases because of the change in default behavior.

fujiaxiang committed:
DOC: moved whatsnew to V1.1.0

fujiaxiang committed:
Merge remote-tracking branch 'upstream/master' into series_rolling_count_ignores_min_periods

fujiaxiang committed:
Merge remote-tracking branch 'upstream/master' into series_rolling_count_ignores_min_periods

fujiaxiang committed:
Merge remote-tracking branch 'upstream/master' into series_rolling_count_ignores_min_periods

WillAyd commented on pandas/core/window/rolling.py:
```diff
@@ -1186,13 +1186,21 @@ def count(self):
+        # We set the default value min_periods to be 0 because count method
+        # is meant to count NAs, we don't want it by default requires all
+        # values in the window to be valid to produce a valid count
+        min_periods = 0 if self.min_periods is None else self.min_periods
```
Is this case tested? Still a little unclear why this is required; does the change you have below to just pass `min_periods` as an argument not fix the issue?

fujiaxiang replied:
This was not explicitly tested, but are partially covered by some existing test cases.
I have added explicit tests.

See below my comment for explanation on why I think this should be the behavior.

fujiaxiang replied:
wait I think I may have misinterpreted some of the existing code.
give me another couple of hours I will verify it.
hopefully I can make this API entirely consistent with others

fujiaxiang replied:
I realize my earlier analysis was incorrect. We can in fact change the behavior of `rolling.count` and `expanding.count` to be consistent with other APIs. Have pushed the code change!

jreback reviewed (changes requested):
ok with this for 1.0.0 if you can move this note

fujiaxiang committed:
TST: added more tests (GH26996)

fujiaxiang committed:
Merge branch 'master' into series_rolling_count_ignores_min_periods

fujiaxiang committed:
Merge remote-tracking branch 'upstream/master' into series_rolling_count_ignores_min_periods

fujiaxiang committed:
BUG: updated rolling and expanding count for consistency (GH26996)

Updated the behavior of rolling and expanding count so that it becomes consistent with all other rolling and expanding functions.
Also updated many test cases to reflect this change of behavior.

fujiaxiang commented:
After further investigation, I realize we can in fact make `count` function behave consistently with other `rolling` and `expanding` functions.

I have pushed the code changes, and also updated many test cases that were "incorrect".

Now the `count` function correctly handles ``min_periods`` argument, which was ignored previously.

WillAyd commented on pandas/core/window/rolling.py:
```diff
@@ -1186,13 +1186,18 @@ def count(self):
 
+        min_periods = self.min_periods
+        if min_periods is not None and not self.center:
+            # this is required as window is mutated above
```
Can you clarify a bit further? I don't think we do this type of check in other functions like `cov` - just curious why we would have to do here

fujiaxiang replied:
> I think looks good but given scope of change I don't think this is suitable for 1.0 - it isn't backwards compatible in its current form with the change to how count works, so probably best for 1.1

@WillAyd Would you think it's a good idea if we roll-out the previous commit which is backward-compatible for V1.0.0, then another PR to roll-out this long term fix for V1.1?

I will probably further look if we can clean this part too for V1.1
```python
window = min(window, len(obj)) if not self.center else window

...

min_periods = self.min_periods
        if min_periods is not None and not self.center:
            # this is required as window is mutated above
            min_periods = min(min_periods, window)
```

fujiaxiang replied:
@WillAyd You are right, after rounds of code cleaning, this part `window = min(window, len(obj)) if not self.center else window` is no longer needed, so the part you commented is also not needed any more.

WillAyd reviewed (commented):
I think looks good but given scope of change I don't think this is suitable for 1.0 - it isn't backwards compatible in its current form with the change to how count works, so probably best for 1.1 @jreback ?

fujiaxiang committed:
CLN: further cleaned code (GH26996)

fujiaxiang commented:
After a few rounds of debugging and code clean-up, now `rolling.count` behaves completely consistent with other APIs (e.g `rolling.mean`).

@jreback mentioned the fix is ok for V1.0 at 2 commits ago (bfe10f08696d0073db27bac898d6629fbef09f5d), @WillAyd thinks the scope of change is too large for V1.0 since last commit (c0046b6f55c68ffc9d63bf8624fc1e926a4b00df).

Would it be a good idea to roll-out the backward-compatible smaller fix at 2 commits ago for V1.0, and the longer-term complete fix for V1.1 with a separate PR?

fujiaxiang commented:
test this pleaae

fujiaxiang commented:
reset this please

fujiaxiang commented:
Is there a way to retrigger checks without new commits? Recently CI seems to randomly fail a lot.

fujiaxiang commented:
reset

TomAugspurger commented:
If you merge master & repush then the CI should be fixed.

fujiaxiang committed:
Merge remote-tracking branch 'upstream/master' into series_rolling_count_ignores_min_periods

fujiaxiang commented:
@TomAugspurger 
That's what I have been trying to do, but it doesn't always work. Sometimes I have to do that a few times before the checks go green. The problem is, I will have to wait for the master to have new commits and doing that a few times may cost me one or two days. Is there anyway I can reliably re-trigger checks without having to push new commits (including merge master)?

jreback commented:
to retrigger a build if u think it’s actually an intermittent issue (normally pretty rare)
you can close and open the PR

jreback commented:
@mroeschke if you'd have a look here

mroeschke commented on pandas/tests/window/test_rolling.py:
```diff
@@ -446,3 +446,38 @@ def test_min_periods1():
     tm.assert_series_equal(result, expected)
+
+
+@pytest.mark.parametrize("test_series", [True, False])
```
Can you just parameterize over the constructor and use `tm.assert_equal` at the end?

fujiaxiang replied:
done

mroeschke commented on pandas/tests/window/test_rolling.py:
```diff
@@ -446,3 +446,38 @@ def test_min_periods1():
+        tm.assert_frame_equal(result, expected)
+
+
+@pytest.mark.parametrize("test_series", [True, False])
```
Same

fujiaxiang replied:
done

mroeschke reviewed (commented):
I suspect that originally that `min_periods` was intentionally set `0` to make `count` consistent with the non-rolling `count` methods (i.e. always count all the non-NA values), but overall I think this is an okay change since `rolling` has the ability to handle a `min_periods` concept and the default as you mentioned is set to `0`. 

Just some small testing comments.

fujiaxiang committed:
TST: cleaned up tests (GH26996)

fujiaxiang commented:
> I suspect that originally that `min_periods` was intentionally set `0` to make `count` consistent with the non-rolling `count` methods (i.e. always count all the non-NA values), but overall I think this is an okay change since `rolling` has the ability to handle a `min_periods` concept and the default as you mentioned is set to `0`.
> 
> Just some small testing comments.

@mroeschke Note that we have changed the default value of ``min_periods`` in the latest commits. Now it defaults to same value as ``window``, same as other APIs. It was still 0 a few commits ago (bfe10f08696d0073db27bac898d6629fbef09f5d).

mroeschke commented:
Ah thanks for highlighting that @fujiaxiang. Sorry for not totally following the entire conversation to date.

I have a slight preference on keeping the default `min_periods` for `count` to be `0` instead of `window` so `count` is consistent with `count` functions on other pandas objects (e.g. `DataFrame.count`.

Otherwise, I do agree with @WillAyd and think changing the default `min_periods` is a bit aggressive to include 1.0.

fujiaxiang commented:
@mroeschke Would it be a good idea to roll out commit (bfe10f08696d0073db27bac898d6629fbef09f5d) for V1.0, and then separately roll out latest commit for V1.1?

mroeschke commented:
I think we would still need some sort of `DeprecationWarning` on `count` or `min_periods` (although the parameter wasn't working before) since the default behavior of the function would change with the latest commit you have.

Another reason why it may just be simpler to keep the default `min_periods` to 0 and have the PR be a bug fix that `min_periods` now is respected.

fujiaxiang committed:
BUG: changed min_periods default to 0 for rolling and expanding (GH26996)

fujiaxiang committed:
BUG: reverted non-relevant changes accidentally added (GH26996)

fujiaxiang commented:
@mroeschke I have changed default ``min_periods`` back to 0. Could you review?

I have kept most of the changes in the test cases, which now explicitly specifies `min_periods` (`min_periods=0` mostly, `min_periods=min_periods` otherwise where the effect is the same). This would make it easier when we decide to go-ahead with changing the default behavior in future versions.

Should I also create a new issue to add the `DeprecationWarning` and the actual change for next version?

fujiaxiang commented:
> Ah thanks for highlighting that @fujiaxiang. Sorry for not totally following the entire conversation to date.
> 
> I have a slight preference on keeping the default `min_periods` for `count` to be `0` instead of `window` so `count` is consistent with `count` functions on other pandas objects (e.g. `DataFrame.count`.
> 
> Otherwise, I do agree with @WillAyd and think changing the default `min_periods` is a bit aggressive to include 1.0.

@mroeschke btw changing the default ``min_periods`` doesn't make the result that different from `DataFrame.count`. Unlike `rolling.mean` or `rolling.sum` etc, the `min_periods` argument for `rolling.count` only takes effect on the window size, and doesn't care how many NA values are in the window. If we change its default value to the same as `window`, same as other APIs, we would only see NAs in the first few rows (which would be obvious to user that this is caused by `min_periods`), and the rest are consistent.

mroeschke commented:
Yes please create a new issue for possibly changing the default behavior of `rolling.count` with respect to `min_periods`. Good to discuss the change outside this PR.

mroeschke commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -1132,6 +1132,7 @@ Groupby/resample/rolling
 - Bug in :meth:`DataFrame.groupby` when using nunique on axis=1 (:issue:`30253`)
 - Bug in :meth:`GroupBy.quantile` with multiple list-like q value and integer column names (:issue:`30289`)
 - Bug in :meth:`GroupBy.pct_change` and :meth:`core.groupby.SeriesGroupBy.pct_change` causes ``TypeError`` when ``fill_method`` is ``None`` (:issue:`30463`)
+- Bug in :meth:`Rolling.count` and :meth:`Expanding.count` argument ``min_periods`` ignored (:issue:`26996`)
```
``` argument  where ``min_periods`` was ignored ```

fujiaxiang replied:
updated

mroeschke commented:
One small whatsnew comment otherwise LGTM

fujiaxiang committed:
BUG: small change in whatsnew (GH26996)

fujiaxiang commented:
> Yes please create a new issue for possibly changing the default behavior of `rolling.count` with respect to `min_periods`. Good to discuss the change outside this PR.

Created new issue #31302. Let me know what you guys think there.

jreback commented on pandas/core/window/rolling.py:
```diff
@@ -1182,17 +1182,13 @@ class _Rolling_and_Expanding(_Rolling):
-                window=window,
-                min_periods=0,
+                window=self._get_window(),
+                min_periods=self.min_periods if self.min_periods else 0,
```
``self.min_periods or 0``

fujiaxiang replied:
yeah that'd be cleaner, thanks

jreback commented:
does this fully close #26996 ?

jreback reviewed (changes requested):
lgtm. @mroeschke this is a straight bug fix now?

fujiaxiang committed:
CLN: slight cleanup (GH26996)

jreback commented:
thanks for the patch @fujiaxiang

jreback commented:
hmm, this was supposed to backport. very odd. @fujiaxiang can you push a PR to the 1.0.x branch as well (the bot is supposed to do this).

fujiaxiang commented:
PR created (#31319). It has some conflict with target branch in .travis.yml, @jreback could you help resolve?
